apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/KtorWebSocketEngine.kt

```diff
@@ -1,8 +1,53 @@
 package com.apollographql.apollo3.network.ws
 
+import com.apollographql.apollo3.internal.ChannelWrapper
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.channels.Channel
+import okio.ByteString
+import org.khronos.webgl.ArrayBuffer
+import org.khronos.webgl.Uint8Array
+import org.w3c.dom.MessageEvent
+import org.w3c.dom.WebSocket
+import org.w3c.files.Blob
+
 actual class DefaultWebSocketEngine : WebSocketEngine {
-  override suspend fun open(url: String, headers: Map<String, String>): WebSocketConnection {
-    TODO("WebSocket are not supported on JS")
+  override suspend fun open(
+      url: String,
+      headers: Map<String, String>,
+  ): WebSocketConnection {
+    val messageChannel = ChannelWrapper(Channel<String>(Channel.UNLIMITED))
+    val socket = WebSocket(url)
+    val webSocketOpenResult = CompletableDeferred<Unit>()
+    socket.onopen = { webSocketOpenResult.complete(Unit) }
+    socket.onmessage = { messageEvent: MessageEvent ->
+      val data = messageEvent.data
+      if (data is String) {
+        messageChannel.trySend(data)
+      }
+      if (data is ByteString) {
```

Comment:
This is the big thing that I'm not 100% certain about. The type of data is `Any?` so I'm not sure if we will be getting. I know that at least in some cases (the server I was using to verify that this worked) resulted in strings over the socket, but I'm not certain what format a binary message would come back as. I guessed something that could maybe be interpreted as a `ByteString`, but I really don't know, it could be `ByteArray` or `Uint8Array` or something completely different. If anyone has a way to determine it or know of a better way of solving the problem of receiving the messages that would be wonderful.

Reply:
I _think_ `ByteArray` is more likely than `ByteString`. `ByteString` being an okio thing, I don't see how it could be part of an `org.w3c` API. 

According to the [W3C spec](https://websockets.spec.whatwg.org//#feedback-from-the-protocol), `data` can be a `Blob` or an `ArrayBuffer`, not 100% sure how that translates to Kotlin